src/assembly_scanner.cpp

```diff
@@ -0,0 +1,324 @@
+    if (m_fragments != nullptr) {
+        return m_fragments;
+    }
+
+    m_fragments = new std::vector<NONNULL_PTR(std::vector<instruction>)>();
+
+    // We need to do two passes: one to resolve labels, and one to assign targets to jumps. During the first pass, the
+    // fragments are actually constructed. However, jumps may not have valid targets yet, so we need some way to store
+    // the label's name inside a pair<size_t, size_t>. This code relies on the following assumption:
+    static_assert(
+        sizeof(NONNULL_PTR(const string)) <= 2 * sizeof(size_t), "Cannot fit string pointer inside pair of size_t"
+    );
+    // Using an ordered set over any other container so that references are not invalidated after insertion
+    std::set<string> label_names;
+
+    auto label_to_fake_location = [&](const string& name) -> IP {
+        auto iter = label_names.find(name);
+        if (iter == label_names.end()) {
+            auto p = label_names.insert(name);
+            iter = p.first;
+        }
+        NONNULL_PTR(const string) ptr = &*iter;
+        size_t bottom_half = reinterpret_cast<uintptr_t>(ptr) & SIZE_MAX;
+        size_t top_half =
+#if UINTPTR_MAX > SIZE_MAX
+            reinterpret_cast<uintptr_t>(ptr) >> (8 * sizeof(size_t))
+#else
+            SIZE_C(0)
+#endif
+            ;
+        return { top_half, bottom_half };
+    };
+
+    // First pass
+    for (string curr_line; std::getline(*m_program, curr_line);) {
+        // Unquoted semicolons are comments. Remove them.
+        size_t i;
+        for (i = 0; i < curr_line.size(); ++i) {
+            if (curr_line[i] != ';') {
+                continue;
+            }
+            if (i == 0 || curr_line[i - 1] != '\'') {
+                break;
+            }
+        }
+        if (i < curr_line.size()) {
+            curr_line.erase(i);
+        }
+        // If the line is only a comment, move on
+        if (curr_line.empty()) {
+            continue;
+        }
+        // Remove trailing whitespace. If there's only whitespace, skip this line
+        i = curr_line.find_last_not_of(WHITESPACE);
+        if (i == string::npos) {
+            continue;
+        } else {
+            curr_line.erase(i + 1);
+        }
+
+        // Look for labels (non-whitespace in the first column)
+        i = curr_line.find_first_not_of(WHITESPACE ":");
+        assert(i != string::npos);
+        if (i == 0) {
+            // Label, find end
+            i = curr_line.find_first_of(WHITESPACE ":");
+            if (i == string::npos) {
+                i = curr_line.size();
+            }
+            string label = curr_line.substr(0, i);
+
+            DISCARD label_names.insert(label);
+            // Add a NOP for the label to point to
+            add_instruction({ instruction::operation::NOP, instruction::argument() });
+            auto p = m_label_locations.insert({ label, get_current_location() });
```

Comment:
This feels like a bit of a hack, but I'm not sure how else to handle labels at the end of a file. (Nevermind the fact that without this, labels point one instruction too early.)

Reply:
Wait, labels can't appear at the end of a file anyways, because there'd be no instruction to jump to